Improve write performance

valentin-pinkau opened the pull request:
This PR improves the write performance of row-major input data.

Before, the input data was converted by python to column-major and then passed to the rust library.
This conversion was done using `np.asfortranarray`. It turns out that this is unnecessarily slow.

In order to circumvent this issue, the python library determines whether the data is row or column-major and passes this information down to the rust lib.
The data is then converted on the fly using two buffers. 
This way, column-major order is written in any case.

The first buffer contains the data to write in the same order as the input data.
The second buffer is used only if the input data is in row-major and holds the column-major version of the first buffer.

The write performance for 1 Gibibyte on my laptop:
Before:
Write data column-major: 1.14s
Write data row-major: 46.1s

After:
Write data column-major: 1.23s
Write data row-major: 1.11s

Closes #38

philippotto commented on rust/src/mat.rs:
```diff
@@ -1,93 +1,196 @@
-        let dst_off_z = (self.shape.x as usize * self.shape.y as usize * self.voxel_size) as isize;
+        let width = src_box.width();
+        // unified has fast to slow moving indices
+        let unified_width = match self.is_fortran_order {
```
It seems quite weird to me to use `match` on a boolean, when you could also write `let unified_width = if self.is_fortran_order { ... } else { ... }` (shaves off the `true`/`false` literals). But I don't know jack about idiomatic rust, so I might be wrong here :smile: Same in the following blocks..

philippotto commented on rust/src/mat.rs:
```diff
@@ -1,93 +1,196 @@
+                z: self.shape.x,
+            },
+        };
+        let unified_src_shape = match self.is_fortran_order {
```
Maybe refactor this by writing something like `unified_src_shape = transpose_if(src.shape, is_fortran_order)` function? Then, lines 136 to 159 could be expressed in three lines (plus the definition of `transpose_if`).

philippotto commented on rust/src/mat.rs:
```diff
@@ -1,93 +1,196 @@
+                    let column_major_index = x * column_major_stride[0]
+                        + y * column_major_stride[1]
+                        + z * column_major_stride[2];
+                    for byte_offset in 0..self.voxel_size {
```
https://gist.github.com/anonymous/1f34331b2cae6ba9e624c5f9f4f2a458

philippotto commented on python/tests/test_wkw.py:
```diff
@@ -163,6 +163,36 @@ def test_compress():
+    data_shape = (17, 1, 4)
+    data = generate_test_data(np.uint8, data_shape)
+    print(data)
+    with wkw.Dataset.create("tests/tmp", wkw.Header(np.uint8)) as dataset:
```
Did you also test for other dtypes than `np.uint8` and multi-channel data? Maybe add a test case for np.uint16 with two channels or something like that.

amotta commented:
These numbers are insane! Thank you very much for this pull request, Valentin!

Unfortunately, I won't have time to review before tomorrow / early next week. I hope this is okay with you.

valentin-pinkau commented:
@philippotto thank you for your review.
I implemented your remarks and added more tests.
The row-major to column-major conversion gets much more complicated with multiple channels, as the data to be written is expected to be in channel first format. This is why this PR implements a fallback to the usage of `np.asfortranarray`  if there is more than one channel.

amotta commented on rust/src/file.rs:
```diff
@@ -149,7 +147,19 @@ impl File {
+        )?;
+
+        // build second buffer
+        let mut fortran_conversion_buf = vec![0u8; self.header.block_size()];
```
All these buffers start to get a bit confusing (not your fault): There's `self.block_buf` for compression, `buf_mat`, and now `fortran_conversion_buf`. What do you think about using a more consistent naming scheme, like, `src_block_buf` for `buf_mat`, `raw_disk_block_buf` for `fortran_conversion_mat`, and `disk_block_buf` for `block_buf`?

Also, all these buffers are currently allocated and initialized for each visited WKW file. The initialization with zeroes is not really necessary and might slow things down. That's fine for now. But at one point we should profile this and decide whether i) only the buffers that are really needed should be allocated and initialized, ii) buffers should be re-used across reads from / writes to different WKW files within a dataset, and iii) we can (unsafely) skip the zero-initialization.

amotta commented on rust/src/file.rs:
```diff
@@ -177,8 +188,13 @@ impl File {
-            self.write_block(buf_mat.as_slice())?;
+            // write in fortran order
+            if buf_mat.get_is_fortran_order() {
+                self.write_block(buf_mat.as_slice())?;
```
Maybe factor out the `self.write_block` call.

amotta commented on rust/src/mat.rs:
```diff
@@ -1,93 +1,201 @@
     pub voxel_size: usize,
-    pub voxel_type: VoxelType
+    pub voxel_type: VoxelType,
+    is_fortran_order: bool,
```
Can we make this `pub` and get rid of the getter? Not because it's bad right now, but just for consistency.

amotta commented on rust/src/mat.rs:
```diff
@@ -1,93 +1,201 @@
+    }
 
-        offset_vx * self.voxel_size
+    pub fn copy_as_fortran_order(&self, buffer: &mut Mat) -> Result<()> {
```
Hmm. Would it be possible to make this function "private" and simply call it in `copy_from` when appropriate? Seems confusing to have multiple copy methods.

valentin-pinkau replied:
When I move the Fortran conversion into the copy_from, I might have to allocate the `raw_disk_block_buf` for every copy_from call.

amotta commented on rust/src/mat.rs:
```diff
@@ -1,93 +1,201 @@
+            return Err("Writing out of bounds");
+        }
+        if self.is_fortran_order != src.get_is_fortran_order() {
+            return Err("source and destination has to be the same order");
```
Start error message with upper case letter for consistency.

amotta commented on rust/src/mat.rs:
```diff
@@ -1,93 +1,201 @@
+        let unified_width = if self.is_fortran_order {
+            width
+        } else {
+            Vec3 {
```
Maybe add a `Vec3::flip_dimensions` function?

amotta commented on rust/src/mat.rs:
```diff
@@ -1,93 +1,201 @@
-        if !(dst_pos + src_box.width() < (self.shape + 1)) { return Err("Writing out of bounds"); }
+        let width = src_box.width();
+        // unified has fast to slow moving indices
+        let unified_width = if self.is_fortran_order {
```
Please use `len` or `length` instead of `width` for consistency.

valentin-pinkau commented on c/src/lib.rs:
```diff
@@ -224,28 +226,38 @@ pub extern fn dataset_read(
+        let header = dataset.header();
+        let num_channels = header.voxel_type.size() / header.voxel_size as usize;
+        if num_channels > 1 {
+            assert!(is_fortran_order);
```
I moved the check into `Dataset::write_mat`. As the multi-channel data works as before, I think it is fair to implement the fast write support in another PR.

valentin-pinkau commented:
@amotta Thank you for your review!
I addressed most of your remarks. I would like to postpone the multi-channel implementation to another PR.
Reading is quite fast at the moment, as numpy handles the order transparently. When it turns out to slow things down I will have a look :)
